src/EditorFeatures/Core/Tagging/AsynchronousTaggerProvider.cs

```diff
@@ -3,79 +3,189 @@
-    /// the tags is handled by a provided <see cref="IAsynchronousTaggerDataSource{TTag}"/>.  The
-    /// <see cref="AsynchronousTaggerProvider{TTag}"/> handles the jobs of scheduling when to 
-    /// compute tags, managing the collection of tags, and notifying the and keeping the user 
-    /// interface up to date with the latest tags produced.
+    /// Base type of all asynchronous tagger providers (<see cref="ITaggerProvider"/> and <see cref="IViewTaggerProvider"/>). 
     /// </summary>
-    internal class AsynchronousTaggerProvider<TTag> : ITaggerProvider
+    internal abstract partial class AsynchronousTaggerProvider<TTag> : IAsynchronousTaggerDataSource<TTag>, ITaggerProvider, IViewTaggerProvider
         where TTag : ITag
     {
-        private readonly AsynchronousTaggerProviderImpl _taggerImplementation;
+        private readonly object uniqueKey = new object();
+        private readonly IAsynchronousOperationListener asyncListener;
+        private readonly IForegroundNotificationService notificationService;
 
-        /// <summary>
-        /// Creates a new <see cref="AsynchronousTaggerProvider{TTag}"/> using the provided 
-        /// <see cref="IAsynchronousTaggerDataSource{TTag}"/> to determine when to compute tags 
-        /// and to produce tags when appropriate.
-        /// </summary>
-        public AsynchronousTaggerProvider(IAsynchronousTaggerDataSource<TTag> dataSource)
-            : this(null, null, dataSource)
+        public virtual TaggerTextChangeBehavior TextChangeBehavior => TaggerTextChangeBehavior.None;
+        public virtual TaggerCaretChangeBehavior CaretChangeBehavior => TaggerCaretChangeBehavior.None;
+        public virtual SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
+
+        public virtual IEqualityComparer<TTag> TagComparer => null;
+
+        public virtual IEnumerable<Option<bool>> Options => null;
+        public virtual IEnumerable<PerLanguageOption<bool>> PerLanguageOptions => null;
+
+        protected AsynchronousTaggerProvider(
+            IAsynchronousOperationListener asyncListener,
+            IForegroundNotificationService notificationService)
         {
+            this.asyncListener = asyncListener;
+            this.notificationService = notificationService;
         }
 
-        internal AsynchronousTaggerProvider(
-            IAsynchronousOperationListener asyncListener,
-            IForegroundNotificationService notificationService,
-            IAsynchronousTaggerDataSource<TTag> dataSource)
+        private TagSource CreateTagSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
         {
-            if (dataSource == null)
+            var options = this.Options ?? SpecializedCollections.EmptyEnumerable<Option<bool>>();
+            var perLanguageOptions = this.PerLanguageOptions ?? SpecializedCollections.EmptyEnumerable<PerLanguageOption<bool>>();
+
+            if (options.Any(option => !subjectBuffer.GetOption(option)) ||
```

Comment:
Shouldn't these still create a tag source but just not ask for tags? What happens if the option changes later?